catalog/resource_grant.go

```diff
@@ -0,0 +1,179 @@
+			Principal: principal,
+			Remove:    permissions.SetToSlice(remove),
+		})
+	}
+	// so that we can deterministic tests
+	sort.Slice(diff, func(i, j int) bool {
+		return diff[i].Principal < diff[j].Principal
+	})
+	return diff
+}
+
+// replacePermissionsForPrincipal merges removal diff of existing permissions on the platform
+func replacePermissionsForPrincipal(a permissions.UnityCatalogPermissionsAPI, securable string, name string, principal string, list catalog.PermissionsList) error {
+	securableType := permissions.Mappings.GetSecurableType(securable)
+	existing, err := a.GetPermissions(securableType, name)
+	if err != nil {
+		return err
+	}
+	err = a.UpdatePermissions(securableType, name, diffPermissionsForPrincipal(principal, list, *existing))
+	if err != nil {
+		return err
+	}
+	return a.WaitForUpdate(1*time.Minute, securableType, name, list, func(current *catalog.PermissionsList, desired catalog.PermissionsList) []catalog.PermissionsChange {
+		return diffPermissionsForPrincipal(principal, desired, *current)
+	})
+}
+
+func filterPermissionsForPrincipal(in catalog.PermissionsList, principal string) (out catalog.PermissionsList) {
+
+	for _, v := range in.PrivilegeAssignments {
+		if v.Principal == principal {
+			out.PrivilegeAssignments = append(out.PrivilegeAssignments, v)
+		}
+	}
+	return
+}
+
+func ResourceGrant() *schema.Resource {
+	s := common.StructToSchema(permissions.UnityCatalogPrivilegeAssignment{},
+		func(m map[string]*schema.Schema) map[string]*schema.Schema {
+
+			m["principal"].ForceNew = true
+
+			allFields := []string{}
+			for field := range permissions.Mappings {
+				allFields = append(allFields, field)
+			}
+			for field := range permissions.Mappings {
+				m[field] = &schema.Schema{
+					Type:          schema.TypeString,
+					Optional:      true,
+					ForceNew:      true,
+					AtLeastOneOf:  allFields,
+					ConflictsWith: permissions.SliceWithoutString(allFields, field),
+				}
+			}
+			return m
+		})
+
+	return common.Resource{
+		Schema: s,
+		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			principal := d.Get("principal").(string)
+			privileges := permissions.ToPrivilegeSlice(d.Get("privileges").(*schema.Set).List())
+			var grants = catalog.PermissionsList{
+				PrivilegeAssignments: []catalog.PrivilegeAssignment{
+					{
+						Principal:  principal,
+						Privileges: privileges,
+					},
+				},
+			}
+			securable, name := permissions.Mappings.KeyValue(d)
+			unityCatalogPermissionsAPI := permissions.NewUnityCatalogPermissionsAPI(ctx, c)
+			err := replacePermissionsForPrincipal(unityCatalogPermissionsAPI, securable, name, principal, grants)
+			if err != nil {
+				return err
+			}
+			d.SetId(fmt.Sprintf("%s/%s", permissions.Mappings.Id(d), principal))
+			return nil
+		},
+		Read: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			principal := d.Get("principal").(string)
```

Comment:
Suggestion: for readability, could we make helper methods to construct/deconstruct the ID? It's not obvious what each part of the ID ought to correspond to.

Reply:
Matched the existing impl for `resource_grants` as closely as I could

Reply:
In any case, added the helper methods to encapsulate the string wrangling